nucliadb_utils/src/nucliadb_utils/nats.py

```diff
@@ -199,42 +218,75 @@ async def reconnected_cb(self):
         logger.warning(
             f"Reconnected to NATS {self._nc.connected_url.netloc}. Attempting to re-subscribe."
         )
-        async with self._lock:
-            existing_subs = self._subscriptions
-            self._subscriptions = []
-            for sub, recon_callback in existing_subs:
-                try:
-                    await sub.drain()
-                    await recon_callback()
-                except Exception:
-                    logger.exception(
-                        f"Error resubscribing to {sub.subject} on {self._nc.connected_url.netloc}"
-                    )
-                    # should force exit here to restart the service
-                    self._healthy = False
-                    raise
-
-            existing_pull_subs = self._pull_subscriptions
-            self._pull_subscriptions = []
-            for pull_sub, task, recon_callback, cancelled in existing_pull_subs:
-                cancelled.set()
-                task.cancel()
-                try:
-                    await task
-                except asyncio.CancelledError:
-                    pass
-                try:
-                    await pull_sub.unsubscribe()
-                    await recon_callback()
-                except Exception:
-                    logger.exception(
-                        f"Error resubscribing to pull subscription on {self._nc.connected_url.netloc}"
-                    )
-                    # should force exit here to restart the service
-                    self._healthy = False
-                    raise
-        self._healthy = True
-        self._last_unhealthy = None  # reset the last unhealthy time
+        # This callback may be interrupted by NATS. In order to avoid being interrupted, we spawn our own task to do
+        # the reconnection. If we are interrupted we could lose subscriptions
+        self._needs_reconnection = True
+        if self._reconnect_task is None:
+            self._reconnect_task = asyncio.create_task(self._reconnect())
```

Comment:
I know it doesn't make much difference, but we could await the task here, so reconnected_cb ends up in the proper reconnected state (just for the mental model)